### DISCUSSION TITLE
MIME type for geany projects

### DISCUSSION THREAD

**Discussion by Kycilak** [geany/geany#3374] [Ideas]
I think geany projects should have MIME type other than text/plain so that it can get another default app (geany) than plain text file (eg. text editor). If there is another way to do so, please tell me.

**Comment by Kycilak** [comment]
It should be relatively easy. I have done so by creating file `geany.xml` in `usr/share/mime/packages` as per the instructions in [MIME specs](https://specifications.freedesktop.org/shared-mime-info-spec/latest/ar01s02.html#s2_layout).
The insides of the file:
```
<?xml version="1.0" encoding="UTF-8"?>
<mime-info xmlns="http://www.freedesktop.org/standards/shared-mime-info">
  <mime-type type="text/x-geany-project">
    <comment>Geany project file</comment>
    <glob pattern="*.geany"/>
  </mime-type>
</mime-info>
```
If there is a "proper geany" way to do this, please tell me still. If there is none, I think this is relatively easy solution to be included in geany project.

**Comment by elextr** [comment]
On cinnamon file manager right click any `.geany` file, `Open With`>`Other Application` opens a dialog, select Geany and `Set as Default`.  From now on shows Geany by default for project files.  Done.

Pretty sure most other desktops have similar capability.

Geany itself does not use MIME data to identify filetypes because it is so often incorrect, or at least not precise enough.

A quick read of the linked spec seems to require something to be run after the xml is installed to update the caches.  How common is that application?

**Reply by Kycilak** [reply]
Sure I can set Geany as the default for `.geany` files. The problem is that every other text file then opens in Geany as well. I hypothesized that is due to the `.geany` files having MIME type `text/plain`. This appears to be the reason because with the added MIME type I am able to select Geany as default for `.geany` without it affecting all other `text/plain` files.

**Reply by Kycilak** [reply]
`update-mime-database` is needed after the `.xml` file is introduced. I didn't need to install it, it probably comes preinstalled with Mint (and probably many other flavors).
I have little experience with the internals OS, so I may not know how widespread and reliable MIME types are. But I have found other applications use it. On my PC ocenaudio, audacity, libreoffice and other applications have `.xml` file and I suppose they use it to differentiate their corresponding files by extensions the same way I wanted Geany to do.

**Comment by elextr** [comment]
> On my PC ocenaudio, audacity, libreoffice and other applications have .xml file and I suppose they use it to differentiate their corresponding files by extensions the same way I wanted Geany to do.

The _file manager_ uses it to select the application to run, the applications are unlikely to use it themselves, they know what their file extensions are :-)

But I think the file should be installed in `${prefix}/share/mime/packages` not dive directly to `/usr`.  Geany already installs data to a number of subdirectories of `${prefix}/share` including `geany.desktop` (another freedesktop thing) to `${prefix}/share/applications` so `${prefix}/share` should be in `XDG_DATA_DIRS` already.

If "somebody" made a pull request to add the file and integrate it with the autotools and meson installers (and have them check for and not attempt to run `update-mime-database` if its not present) it would probably be accepted.  Probably should be integrated with the translation system too, I notice xml files from other applications have many translations of the `<comment>` entity.

**Reply by Kycilak** [reply]
I see, I talked about `\usr` because that's what I reached first. I have little knowledge about what the proper structure is. I wasn't concerned as it was a change just on my local machine.
I understand that default application management for files isn't the responsibility of Geany. however Geany is responsible for setting clues for file manager upon installation (if there is to be the functionality described above).
I have not yet contributed much, have little experience with programming and little knowledge of OS. Maybe in a couple of years I can propose the changes myself. In the meantime I've decided to start this discussion to spread the idea or get told it is problematic in some way.
